### DISCUSSION TITLE
process.env is undefined in app directory

### DISCUSSION THREAD

**Discussion by AviAvinav** [vercel/next.js#48151] [Help] [answered]
### Summary

I am trying to create a contact form with emailjs in the experimental app directory.

This is my onSumbit function for the form:

```tsx
const onSubmit = (e: FormEvent) => {
    e.preventDefault();
    emailjs
      .sendForm(
        process.env.SERVICE_ID,
        process.env.TEMPLATE_ID,
        form.current || "",    // 'form' is a RefObject from useRef for the form.
        process.env.PUBLIC_KEY
      )
      .then(
        (result) => {
          toast("Message sent successfully!");
        },
        (error) => {
          toast("Message failed to send.");
        }
      );
  };
```

This is inside the file `app/form.tsx`, this file also contains the UI component of the form.

I keep getting the error: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.

The function seems to work and send emails when I hardcode the keys (which I can't do in production) but stops when due to the above mentioned error with process.env

How can I fix this?

**Comment by AviAvinav** [comment, marked as answer]
In case anyone else has this same issue, I just had to expose the variables to the client side by prefixing them with `NEXT_PUBLIC_`. For example, `SERVICE_ID` to `NEXT_PUBLIC_SERVICE_ID`.

**Reply by henriquedevelops** [reply]
I don't get it

What is the point of using an environment variable if you are going to make it public anyway?

**Reply by saurabhje** [reply]
Non-`Next_PUBLIC` environment variables are available only for the server components, using `NEXT_PUBLIC` will expose them to client JavaScript bundle. Hence the sensitive variables are often accessed in the serve components. We use `NEXT_PUBLIC` for non-sensitive variables which we require to make subsequent and frequent requests. 

You can read more about this here - https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser
